### PR TITLE
Validate network interface name

### PR DESCRIPTION
**Pull request by rhrazdil**
Virtual Machine with Network interface that contains "." in its name fails to boot
despite k8s allowing "." in object names. Adding a validation to refuse VMs with
NICs containing "." in their name.

Signed-off-by: Radim Hrazdil <rhrazdil@redhat.com>

**What this PR does / why we need it**:
PR fixes issue with VM not booting up due to a network interface containing dot (".") in its name.

**Special notes for your reviewer**:

**Release note**:
```release-note
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1853911
```

### CODE REVIEW

**Review by phoracek — COMMENTED** [review]
Why exactly it fails to boot? What is getting broken by the dot? Is it Kubernetes, Multus, our binding mechanism or something else?

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go [hunk 1]
```diff
@@ -623,6 +623,15 @@ func ValidateVirtualMachineInstanceSpec(field *k8sfield.Path, spec *v1.VirtualMa
 			})
 		}
 
+		// Check interface name doesn't contain dots
+		if strings.Contains(iface.Name, ".") {
+			causes = append(causes, metav1.StatusCause{
+				Type:    metav1.CauseTypeFieldValueInvalid,
+				Message: fmt.Sprintf("Network interface name cannot contain \".\": %s", iface.Name),
```
Comment (phoracek) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:630]: Once we see what is the cause of this issue, we can hopefully define what is the format of a valid network and if dot is the only invalid character.

**Comment by rhrazdil** [comment]
After some debugging I found with @phoracek  help that the problem is either in libvirt [`DomainDefineXML`](https://github.com/kubevirt/kubevirt/blob/906d2e1a84749993d1a716908bba8142eb58bca4/vendor/libvirt.org/libvirt-go/connect.go#L933) or [`GetXMLDesc`](https://github.com/kubevirt/kubevirt/blob/906d2e1a84749993d1a716908bba8142eb58bca4/vendor/libvirt.org/libvirt-go/domain.go#L1168).

I learned that the interface name is also used in environment variable names (`KUBEVIRT_RESOURCE_NAME_<nic name>`) passed to virt-launcher pod, and as such need to conform to format for env variables names, which is more restrictive with the following validation regex `[-._a-zA-Z][-._a-zA-Z0-9]*`.
From this, the dot (`.`) is the only character that causes issues, so I suggest using the following regex for interface name validation: '^[-._a-zA-Z][-._a-zA-Z0-9]*$':

Petr WDYT?

**Comment by phoracek** [comment]
@rhrazdil thanks for looking into it.

Regarding the regex, it sounds reasonable to restrict the allowed characters. Your regex however allows for a dot, we want to prevent that, right? And is there a reason not to allow it starting with a number? The environment variable is already prefixed.

**Comment by rhrazdil** [comment]
> Your regex however allows for a dot

Right, forgot to remove it after copy-pase...

> And is there a reason not to allow it starting with a number? The environment variable is already prefixed.

Good point! `[-_a-zA-Z0-9]*` should be all we need then

**Comment by phoracek** [comment]
Sounds good to me. Just make sure to explain what is allowed in human language and not by just showing the regex.

Also, please add a unit test for this.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go [hunk 2]
```diff
@@ -623,6 +623,16 @@ func ValidateVirtualMachineInstanceSpec(field *k8sfield.Path, spec *v1.VirtualMa
 			})
 		}
 
+		// Check if the interface name has correct format
+		isValid := regexp.MustCompile(`^[A-Za-z0-9-_]+$`).MatchString
+		if !isValid(iface.Name) {
+			causes = append(causes, metav1.StatusCause{
+				Type:    metav1.CauseTypeFieldValueInvalid,
+				Message: "Network interface name must be made up of following characters [A-Za-z0-9-_]",
```
Comment (phoracek) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:631]: This is helpful only to people who understand regex. e.g. "of alphanumeric characters, dashes (-) and underscores (_)"

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go [hunk 3]
```diff
@@ -1574,7 +1584,7 @@ func validateDisks(field *k8sfield.Path, disks []v1.Disk) []metav1.StatusCause {
 
 		// Verify serial number is made up of valid characters for libvirt, if provided
 		isValid := regexp.MustCompile(`^[A-Za-z0-9_.+-]+$`).MatchString
-		if disk.Serial != "" && !isValid(disk.Serial) {
+		if disk.Serial != "" && !isValid(disk.Serial) {;
```
Comment (phoracek) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:1587]: typo

**Review by phoracek — CHANGES_REQUESTED** [review]
Please add a test

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go (hunk 3 above)
Comment (rhrazdil) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go]: ops, fixed

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go (hunk 2 above)
Comment (rhrazdil) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go]: Fixed

**Comment by rhrazdil** [comment]
@phoracek At this moment, with the change in this PR we are also rejecting network interfaces with empty names.
We discussed this a while back. Are you okay with keeping this change in this PR or would it be preferable to introduce separately?

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go [hunk 4]
```diff
@@ -2560,8 +2584,9 @@ var _ = Describe("Function getNumberOfPodInterfaces()", func() {
 				Pod:    &v1.PodNetwork{},
 				Multus: &v1.MultusNetwork{NetworkName: "testnet1"},
 			},
+			Name: "testnet",
 		}
-		iface1 := v1.Interface{Name: net1.Name}
+		iface1 := v1.Interface{Name: "testnet"}
```
Comment (phoracek) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go:2589]: How is this related to the PR?

Comment (phoracek) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go:2589]: (If it is an unrelated fix, I don't mind piggy backing it in this PR, but let's at least give it a dedicated commit)

Comment (rhrazdil) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go:2589]: `^[A-Za-z0-9-_]+$` doesn't allow empty string as interface name

Comment (rhrazdil) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go:2589]: see my comment above ^^

**Comment by phoracek** [comment]
@rhrazdil oh now I remember **that** discussion :D yeah, better to fix that too. That also explains the other change.

/lgtm

Thanks!

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go (hunk 4 above)
Comment (phoracek) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go:2589]: Nevermind. Got it now

**Review by phoracek — APPROVED** [review]

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go (hunk 4 above)
Comment (jean-edouard) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go:2589]: Nit: Now that `net1` has a Name, can't we revert to `net1.Name`?

**Review by jean-edouard — APPROVED** [review]

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go (hunk 4 above)
Comment (rhrazdil) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go]: @jean-edouard Yes, you have a point. Done.

**Review by phoracek — APPROVED** [review]
